sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/IfDirectiveHelper.cs

```diff
@@ -0,0 +1,193 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2018 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace SonarAnalyzer.Helpers
+{
+    internal static class IfDirectiveHelper
+    {
+        /// <summary>
+        /// Returns a list of the names of #if [NAME] sections that the specified
+        /// node is contained in.
+        /// </summary>
+        /// <remarks>
+        /// Note: currently we only handle directives with simple identifiers e.g. #if FOO, #elif FOO
+        /// We don't handle logical operators e.g. #if !DEBUG, and we don't handle cases like
+        /// #if !DEBUG ... #else... :DEBUG must be true in the else case.
+        /// </remarks>
+        public static IEnumerable<string> GetActiveConditionalCompilationSections(SyntaxNode node)
```

Comment:
Helper class to find all of the active directive processor sections before the specified node.

Directive processor syntax is awkward to handle: the keywords are embedded in leading or trailing trivia, which could be attached to a sibling node, not a parent node. Also, _DirectiveSyntaxNodes_ types do exist, but they aren't part of the syntax tree in the normal way (e.g. their _Parent_ property always _null_). It looks like the only way to hold of them is to find the trivia they are embedded in and call _trivia.GetStructure()_.

I ended up using a syntax walker to walk down the tree until reaching the target node. The problem then is that the code in the class code be like this:
```
#if DEBUG
   Console.WriteLine(); // (1) compliant
#endif

Console.WriteLine(); // (2) non-compliant
```

We can't simply stop at the first _#if DEBUG_ and say that we're in a DEBUG block, as then we'd not report on the _Console.WriteLine_ at (2) i.e. we'd have swapped a few false positives for probably a larger number of false negatives.

To be reasonably accurate, we have to keep track of the directive blocks that are currently "open".
The approach I used is:
(1) find all if/else/elif/end blocks above the current node,
(2) walk through the list, pushing onto a stack when a new block opens and popping when a block closes.